File: cdm/cpp/system/equipment/anesthesiamachine/actions/SETubeCuffLeak.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
#include "stdafx.h"
#include "system/equipment/anesthesiamachine/actions/SETubeCuffLeak.h"
#include "properties/SEScalar0To1.h"

SETubeCuffLeak::SETubeCuffLeak() : SEAnesthesiaMachineAction()
{
  m_Severity = nullptr;
}

SETubeCuffLeak::~SETubeCuffLeak()
{
  Clear();
}

void SETubeCuffLeak::Clear()
{
  SEAnesthesiaMachineAction::Clear();
  SAFE_DELETE(m_Severity);
}

bool SETubeCuffLeak::IsValid() const
{
  return SEAnesthesiaMachineAction::IsValid() && HasSeverity();
}

bool SETubeCuffLeak::IsActive() const
{
  return HasSeverity() ? !m_Severity->IsZero() : false;
}

void SETubeCuffLeak::Load(const cdm::TubeCuffLeakData& src, SETubeCuffLeak& dst)
{
  SETubeCuffLeak::Serialize(src, dst);
}
void SETubeCuffLeak::Serialize(const cdm::TubeCuffLeakData& src, SETubeCuffLeak& dst)
{
  SEAnesthesiaMachineAction::Serialize(src.anesthesiamachineaction(), dst);
  if (src.has_severity())
    SEScalar0To1::Load(src.severity(), dst.GetSeverity());
}

cdm::TubeCuffLeakData* SETubeCuffLeak::Unload(const SETubeCuffLeak& src)
{
  cdm::TubeCuffLeakData* dst = new cdm::TubeCuffLeakData();
  SETubeCuffLeak::Serialize(src, *dst);
  return dst;
}
void SETubeCuffLeak::Serialize(const SETubeCuffLeak& src, cdm::TubeCuffLeakData& dst)
{
  SEAnesthesiaMachineAction::Serialize(src, *dst.mutable_anesthesiamachineaction());
  if (src.HasSeverity())
    dst.set_allocated_severity(SEScalar0To1::Unload(*src.m_Severity));
}

bool SETubeCuffLeak::HasSeverity() const
{
  return m_Severity != nullptr;
}
SEScalar0To1& SETubeCuffLeak::GetSeverity()
{
  if (m_Severity == nullptr)
    m_Severity = new SEScalar0To1();
  return *m_Severity;
}

void SETubeCuffLeak::ToString(std::ostream &str) const
{
  str << "Anesthesia Machine Action : Tube Cuff Leak";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tSeverity :"; HasSeverity() ? str << *m_Severity : str << "NaN";
  str << std::flush;
}

File: cdm/java/mil/tatrc/physiology/datamodel/system/equipment/anesthesia/actions/SETubeCuffLeak.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.system.equipment.anesthesia.actions;

import com.kitware.physiology.cdm.AnesthesiaMachineActions.TubeCuffLeakData;

import mil.tatrc.physiology.datamodel.properties.SEScalar0To1;

public class SETubeCuffLeak extends SEAnesthesiaMachineAction
{
  protected SEScalar0To1 severity;
  
  public SETubeCuffLeak()
  {
    severity = null;
  }
  
  public void reset()
  {
    super.reset();
    if (severity != null)
      severity.invalidate();
  }
  
  public boolean isValid()
  {
    return hasSeverity();
  }
  
  public static void load(TubeCuffLeakData src, SETubeCuffLeak dst)
  {
    SEAnesthesiaMachineAction.load(src.getAnesthesiaMachineAction(),dst);
    if(src.hasSeverity())
      SEScalar0To1.load(src.getSeverity(),dst.getSeverity());
  }
  public static TubeCuffLeakData unload(SETubeCuffLeak src)
  {
    TubeCuffLeakData.Builder dst = TubeCuffLeakData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SETubeCuffLeak src, TubeCuffLeakData.Builder dst)
  {
    SEAnesthesiaMachineAction.unload(src, dst.getAnesthesiaMachineActionBuilder());
    if (src.hasSeverity())
      dst.setSeverity(SEScalar0To1.unload(src.severity));
  }
  
  /*
   * Severity
   */
  public boolean hasSeverity()
  {
    return severity == null ? false : severity.isValid();
  }
  public SEScalar0To1 getSeverity()
  {
    if (severity == null)
      severity = new SEScalar0To1();
    return severity;
  }
  
  public String toString()
  {
    if (severity != null)
      return "Tube Cuff Leak"
          + "\n\tSeverity: " + getSeverity();
    else
      return "Action not specified properly";
  }
}
